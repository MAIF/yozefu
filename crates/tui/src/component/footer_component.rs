//! The footer component displays contextual information: the current cluster, shortcuts and the last notifications
use crossterm::event::KeyEvent;

use ratatui::{
    layout::Rect,
    style::{Color, Style, Stylize},
    text::{Line, Span},
    Frame,
};
use tokio::sync::mpsc::UnboundedSender;

use crate::{
    action::{Action, Notification},
    error::TuiError,
};

use super::{Component, ComponentName, Shortcut, State};

#[derive(Default)]
pub struct FooterComponent {
    shortcuts: Vec<Shortcut>,
    main_component: ComponentName,
    state: Vec<ComponentName>,
    notification: Option<Notification>,
    action_tx: Option<UnboundedSender<Action>>,
    ticks: u64,
    show_shortcuts: bool,
}

impl FooterComponent {
    fn generate_shortcuts(&self, _state: &State) -> Vec<Span<'static>> {
        let mut spans = vec![];
        for shortcut in &self.shortcuts {
            spans.push(format!("[{}]", shortcut.key).bold());
            spans.push(format!(":{}   ", shortcut.description).into());
        }

        spans
    }

    pub fn show_shortcuts(&mut self, visible: bool) -> &Self {
        self.show_shortcuts = visible;
        self
    }
}

impl Component for FooterComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<(), TuiError> {
        self.action_tx = Some(tx);
        Ok(())
    }

    fn id(&self) -> ComponentName {
        ComponentName::Footer
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::Shortcuts(s, show) => {
                self.shortcuts = s;
                if show {
                    match self.main_component {
                        ComponentName::TopicsAndRecords => self
                            .shortcuts
                            .push(Shortcut::new("CTRL + O", "Hide topics")),
                        ComponentName::Records => self
                            .shortcuts
                            .push(Shortcut::new("CTRL + O", "Show topics")),
                        _ => (),
                    };
                }
                self.shortcuts.push(Shortcut::new("CTRL + H", "Help"));
                self.shortcuts.push(Shortcut::new("TAB", "Next panel"));
                self.shortcuts.push(Shortcut::new("ESC", "Quit"));
            }
            Action::ViewStack((main_component, views)) => {
                self.main_component = main_component;
                self.state = views;
            }
            Action::Notification(notification) => {
                self.ticks = 0;
                self.notification = Some(notification)
            }
            Action::ResetNotification() => {
                self.notification = None;
                self.ticks = 0;
            }
            Action::Tick => {
                self.ticks += 1;
                if self.ticks > 20 {
                    self.notification = None;
                }
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let mut view_stack = self.state.clone();
        view_stack.dedup();
        view_stack.push(state.focused.clone());
        let mut help = vec![];
        help.push(
            format!(" {} ", state.cluster)
                .black()
                .bold()
                .bg(state.theme.white),
        );
        help.push(" ".into());
        for v in view_stack.iter().enumerate() {
            let colors = match v.0 == view_stack.len() - 1 {
                true => (state.theme.bg_active, state.theme.fg_active),
                false => (state.theme.bg_disabled, state.theme.fg_disabled),
            };
            if v.0 > 0 {
                help.push("—".fg(colors.0));
            }
            let prefix = match v.0 {
                0 if self.main_component == ComponentName::TopicsAndRecords => "◧ ",
                0 if self.main_component == ComponentName::Records => "□ ",
                _ => "",
            };

            help.push(
                format!(" {}{:<8}", prefix, v.1.label())
                    .bg(colors.0)
                    .fg(colors.1)
                    .bold(),
            );
        }

        help.push(Span::from("  "));
        if self.show_shortcuts {
            help.extend(self.generate_shortcuts(state));
        }

        let line = Line::from(help);
        f.render_widget(line, rect);

        if let Some(n) = &self.notification {
            let notification =
                Span::styled(n.message.to_string(), Style::default().italic().not_bold());
            let r = Rect::new(
                rect.width
                    .saturating_sub(u16::try_from(n.message.len().checked_sub(3).unwrap_or(3))?),
                rect.y,
                n.message.len() as u16,
                1,
            );
            let notification = match n.level {
                log::Level::Error => notification.fg(Color::LightRed).underlined(),
                log::Level::Warn => notification.fg(Color::Yellow),
                _ => notification,
            };
            f.render_widget(notification, r);
        }
        Ok(())
    }
}
