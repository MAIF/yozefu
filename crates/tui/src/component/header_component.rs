//! The footer component displays contextual information: the current cluster, shortcuts and the last notifications
use crossterm::event::KeyEvent;

use ratatui::{
    Frame,
    layout::Rect,
    style::{Style, Stylize},
    text::{Line, Span},
};
use tokio::sync::mpsc::UnboundedSender;

use crate::{
    action::{Action, Notification},
    error::TuiError,
};

use super::{Component, ComponentName, State};

#[derive(Default)]
pub struct HeaderComponent {
    main_component: ComponentName,
    state: Vec<ComponentName>,
    notification: Option<Notification>,
    action_tx: Option<UnboundedSender<Action>>,
    ticks: u64,
}

impl Component for HeaderComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) {
        self.action_tx = Some(tx);
    }

    fn id(&self) -> ComponentName {
        ComponentName::Header
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::ViewStack((main_component, views)) => {
                self.main_component = main_component;
                self.state = views;
            }
            Action::Notification(notification) => {
                self.ticks = 0;
                self.notification = Some(notification)
            }
            Action::Tick => {
                self.ticks += 1;
                if self.ticks > 20 {
                    self.notification = None;
                }
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let mut view_stack = self.state.clone();
        view_stack.dedup();
        view_stack.push(state.focused.clone());
        let mut help = vec![];
        help.push(
            format!(" {} ", state.cluster)
                .black()
                .bold()
                .bg(state.theme.white),
        );
        help.push(" ".into());
        for v in view_stack.iter().enumerate() {
            let colors = match v.0 == view_stack.len() - 1 {
                true => (state.theme.bg_active, state.theme.fg_active),
                false => (state.theme.bg_disabled, state.theme.fg_disabled),
            };
            if v.0 > 0 {
                #[cfg(not(debug_assertions))]
                help.push("—".fg(colors.0));
                #[cfg(debug_assertions)]
                help.push("".to_string().bg(colors.0).fg(state.theme.bg));
            }
            let prefix = match v.0 {
                0 if self.main_component == ComponentName::TopicsAndRecords => "◧ ",
                0 if self.main_component == ComponentName::Records => "□ ",
                _ => "",
            };

            help.push(
                format!(" {}{:<8}", prefix, v.1.label())
                    .bg(colors.0)
                    .fg(colors.1)
                    .bold(),
            );
            #[cfg(debug_assertions)]
            help.push("".fg(colors.0));
        }

        help.push(Span::from("  "));

        let line = Line::from(help);
        f.render_widget(line, rect);

        if let Some(n) = &self.notification {
            let notification =
                Span::styled(n.message.to_string(), Style::default().italic().not_bold());
            let r = Rect::new(
                rect.width
                    .saturating_sub(u16::try_from(n.message.len().checked_sub(1).unwrap_or(1))?),
                rect.y,
                n.message.len() as u16,
                1,
            );
            let notification = match n.level {
                tracing::Level::ERROR => notification.fg(state.theme.red).underlined(),
                tracing::Level::WARN => notification.fg(state.theme.yellow),
                _ => notification,
            };
            f.render_widget(notification, r);
        }
        Ok(())
    }
}

#[cfg(test)]
use crate::assert_draw;

#[test]
fn test_draw() {
    let mut component = HeaderComponent::default();
    assert_draw!(component, 50, 3)
}
