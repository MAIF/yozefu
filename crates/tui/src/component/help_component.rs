//! Component showing the help

use crossterm::event::{KeyCode, KeyEvent};
use ratatui::{
    Frame,
    layout::Rect,
    style::Stylize,
    text::{Line, Span},
    widgets::{Block, BorderType, Borders, Clear, Padding, Paragraph, Wrap},
};

use crate::{Action, error::TuiError};

use super::{
    Component, ComponentName, Shortcut, State, issue_component::IssueComponent,
    scroll_state::ScrollState,
};

const HELP_HEIGHT: usize = 42;
const TEN_MINUTES_FRAME: usize = 30 * 60 * 10;
const REPOSITORY_URL: &str = concat!(
    "      https://github.com/MAIF/yozefu/tree/v",
    env!("CARGO_PKG_VERSION")
);

#[derive(Default)]
pub(crate) struct HelpComponent {
    scroll: ScrollState,
    rendered: usize,
}

impl Component for HelpComponent {
    fn id(&self) -> ComponentName {
        ComponentName::Help
    }

    fn shortcuts(&self) -> Vec<Shortcut> {
        vec![]
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        self.rendered = 0;
        match key.code {
            KeyCode::Char('k') | KeyCode::Down => {
                self.scroll.scroll_to_next_line();
            }
            KeyCode::Char('j') | KeyCode::Up => {
                self.scroll.scroll_to_previous_line();
            }
            KeyCode::Char('[') => {
                self.scroll.scroll_to_top();
            }
            KeyCode::Char(']') => {
                self.scroll.scroll_to_bottom();
            }
            _ => (),
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        f.render_widget(Clear, rect);

        let block = Block::default()
            .borders(Borders::ALL)
            .padding(Padding::horizontal(2))
            .border_type(BorderType::Rounded)
            .title(" Help ");

        let block = self.make_block_focused_with_state(state, block);

        let text = vec![
            Line::from(Span::raw("")),
            Line::from(Span::raw("")),
            Line::from("                                                           Key      Description").bold(),
            Line::from("                                                             /      Focus search input"),
            Line::from("                                                           ESC      Close the window/app"),
            Line::from("                                                           TAB      Focus next window"),
            Line::from("                                                   SHIFT + TAB      Focus previous window"),
            Line::from(""),

            Line::from("                                                      Variable      Type                        Alias       Description").bold(),
            Line::from(vec![Span::raw("                                                         topic      "), Span::from("String").fg(state.theme.green), Span::from("                          t").fg(state.theme.blue), Span::from("        Kafka topic")]),
            Line::from(vec![Span::raw("                                                        offset      "), Span::from("Number").fg(state.theme.green), Span::from("                          o").fg(state.theme.blue), Span::from("       Offset of the record")]),
            Line::from(vec![Span::raw("                                                           key      "), Span::from(""), Span::from("                                k").fg(state.theme.blue), Span::from("       Key of the record")]),
            Line::from(vec![Span::raw("                                                         value      "), Span::from(""), Span::from("                                v").fg(state.theme.blue), Span::from("       Value of the record")]),
            Line::from(vec![Span::raw("                                                     partition      "), Span::from("Number").fg(state.theme.green), Span::from("                          p").fg(state.theme.blue), Span::from("       Partition of the record")]),
            Line::from(vec![Span::raw("                                                     timestamp      "), Span::from("String").fg(state.theme.green), Span::from("                         ts").fg(state.theme.blue), Span::from("       Timestamp of the record (RFC 3339) â†’ 2025-06-01T12:00:00.000+02:00")]),
            Line::from(vec![Span::raw("                                                          size      "), Span::from("Number").fg(state.theme.green), Span::from("                         si").fg(state.theme.blue), Span::from("       Size of the record")]),
            Line::from(vec![Span::raw("                                                       headers      "), Span::from("Map<String, String>").fg(state.theme.green), Span::from("             h").fg(state.theme.blue), Span::from("       Headers of the record")]),
            Line::from(Span::raw("")),

            Line::from(vec![Span::from("                                                      Operator"), Span::from("      Type").fg(state.theme.green), Span::from("                                    Description").bold()]),
            Line::from(vec![Span::from("                                     == | != | > | >= | < | <="), Span::from("      Number | String").fg(state.theme.green), Span::from("                         Wayne's world, party time! Excellent!")]),
            Line::from(vec![Span::from("                                                 contains | ~="), Span::from("      String").fg(state.theme.green), Span::from("                                  Test if the variable contains the specified string")]),
            Line::from(vec![Span::from("                                                   starts with"), Span::from("      String").fg(state.theme.green), Span::from("                                  Test if the variable starts with the specified string")]),
            Line::from(""),


            Line::from(vec![Span::from("                                                        Clause      Syntax                                  Description").bold()]),
            Line::from(vec![Span::from("                                                         limit      limit <"), Span::from("number").fg(state.theme.yellow), Span::from(">                          Limit the number of kafka records to receive")]),
            Line::from(vec![Span::from("                                                          from      from <"), Span::from("begin").fg(state.theme.yellow), Span::from("|"), Span::from("end").fg(state.theme.yellow), Span::from("|"), Span::from("date").fg(state.theme.yellow), Span::from("|"), Span::from("offset").fg(state.theme.yellow), Span::from(">            Start consuming records from the beginning, the end or a date")]),
            Line::from(vec![Span::from("                                                      order by      order by <"), Span::from("var").fg(state.theme.yellow), Span::from("> <"), Span::from("asc").fg(state.theme.yellow), Span::from("|"), Span::from("desc").fg(state.theme.yellow), Span::from(">               Sort kafka records")]),
            Line::from(""),

            Line::from("                                                         Input      Description").bold(),
            Line::from(r#"                                    timestamp >= "1 hours ago"      All records published within the last hour"#),
            Line::from(r#"v contains "rust" and partition == 2 from beginning limit 1000      The first 1_000 kafka records from partition 2 containing 'rust' in the value"#),
            Line::from(r#"              (key == "ABC") || (key ~= "XYZ") from end - 5000      Among the latest 5_000 records, return the records where the key is "ABC" or the key contains "XYZ""#),
            Line::from(r#"                      value.hello == "world" order by key desc      Any kafka JSON record with a JSON property "hello" with the value "world", sorted by key in descending order"#),
            Line::from(""),
            Line::from(vec![
                Span::from("                                                         Theme").bold(),
                Span::from(format!(
                                        "      Theme is '{}'. run `yozf config get themes` to list available themes.",
                                        state.theme.name
                                    ))
            ]),
            Line::from(vec![
                Span::from("                                                 Configuration").bold(),
                Span::from(format!("      '{}'", state.configuration_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                          Logs").bold(),
                Span::from(format!("      '{}'", state.logs_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                       Filters").bold(),
                Span::from(format!("      '{}'", state.filters_dir.display()))
            ]),
            Line::from(vec![
                Span::from("                                                        Themes").bold(),
                Span::from(format!("      '{}'", state.themes_file.display()))
            ]),
            Line::from(vec![
                Span::from("                                                       Version").bold(),
                Span::from(REPOSITORY_URL)
            ]),
            Line::from(""),
        ];

        let paragraph = Paragraph::new(text)
            .wrap(Wrap { trim: false })
            .scroll((self.scroll.value(), 0));
        f.render_widget(paragraph.block(block), rect);

        if self.rendered > TEN_MINUTES_FRAME {
            let mut issue = IssueComponent::default();
            issue.draw(f, rect, state)?;
        }

        self.scroll.draw(f, rect, HELP_HEIGHT);
        self.rendered += 1;

        Ok(())
    }
}

#[cfg(test)]
use crate::assert_draw;

#[test]
fn test_draw() {
    let mut component = HelpComponent::default();
    assert_draw!(component, 300, 60)
}
