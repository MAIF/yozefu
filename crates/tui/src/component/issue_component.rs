use ratatui::{
    Frame,
    layout::{Constraint, Flex, Layout, Rect},
    style::{Style, Stylize},
    text::Line,
    widgets::{Block, BorderType, Borders, Clear, Padding, Paragraph},
};

use crate::error::TuiError;

use super::{Component, ComponentName, State};

#[derive(Default)]
pub(crate) struct IssueComponent {}

impl Component for IssueComponent {
    fn id(&self) -> ComponentName {
        ComponentName::Help
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let block = Block::default()
            .borders(Borders::ALL)
            .padding(Padding::symmetric(5, 0))
            .style(Style::default())
            .border_type(BorderType::Rounded);

        let block = self.make_block_focused(state, block);
        let paragraph = Paragraph::new(vec![
            Line::from(""),
            Line::from("ʕノ•ᴥ•ʔノ ︵ ┻━┻"),
            Line::from("Alt: Your mate flipping the desk").italic(),
            Line::from(""),
            Line::from("Are you struggling with the tool?"),
            Line::from("Leave us an issue so we can improve it:"),
            Line::from("https://github.com/MAIF/yozefu/issues").bold(),
            Line::from(""),
            Line::from("Press any key to close.").italic(),
        ]);

        let mut rect = rect;
        rect.width -= 2;
        rect.y += 1;

        let [area] = Layout::horizontal([Constraint::Length(54)])
            .flex(Flex::End)
            .areas(rect);
        let [area] = Layout::vertical([Constraint::Length(11)])
            .flex(Flex::Start)
            .areas(area);

        f.render_widget(Clear, area);
        f.render_widget(paragraph.block(block), area);

        Ok(())
    }
}
