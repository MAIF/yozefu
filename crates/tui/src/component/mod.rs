mod footer_component;
mod help_component;
mod issue_component;
mod progress_bar_component;
mod record_details_component;
mod records_component;
mod root_component;
mod schemas_component;
mod scroll_state;
mod search_component;
mod shortcut;
mod state;
mod topic_details_component;
mod topics_and_records_component;
mod topics_component;
pub mod ui;
mod vertical_scrollable_block;

use crossterm::event::{KeyEvent, MouseEvent};
use ratatui::{
    Frame,
    layout::Rect,
    style::{Style, Stylize},
    widgets::{Block, BorderType},
};
pub(crate) use root_component::RootComponent;
pub(crate) use shortcut::Shortcut;
use strum::Display;
use tokio::sync::mpsc::UnboundedSender;
pub use ui::Ui;

use std::sync::{Arc, LazyLock, Mutex};

pub use state::State;

use serde::Deserialize;

use crate::{Action, TuiError, records_buffer::RecordsBuffer, tui::Event};

pub(crate) type ConcurrentRecordsBuffer = LazyLock<Arc<Mutex<RecordsBuffer>>>;
static BUFFER: ConcurrentRecordsBuffer =
    LazyLock::new(|| Arc::new(Mutex::new(RecordsBuffer::new())));

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub(crate) enum FocusDirection {
    Top,
    Left,
    Right,
    Bottom,
}

#[derive(Debug, Clone, Display, Hash, PartialEq, Eq, Deserialize, PartialOrd, Ord)]
pub(crate) enum ComponentName {
    Records,
    Topics,
    Footer,
    RecordDetails,
    TopicsAndRecords,
    RecordsView,
    TopicDetails,
    Main,
    Search,
    Dialog,
    Help,
    Schemas,
}

impl ComponentName {
    pub fn label(&self) -> String {
        match &self {
            ComponentName::RecordDetails => "Record".to_string(),
            ComponentName::TopicDetails => "Topic".to_string(),
            _ => self.to_string(),
        }
    }
}

impl Default for ComponentName {
    fn default() -> Self {
        Self::Topics
    }
}

pub(crate) trait WithHeight: Component {
    fn content_height(&self) -> usize {
        0
    }
}

pub(crate) trait Component {
    fn register_action_handler(&mut self, _tx: UnboundedSender<Action>) {}

    fn id(&self) -> ComponentName;

    fn make_block_focused_with_state<'a>(&self, state: &State, block: Block<'a>) -> Block<'a> {
        match state.focused == self.id() {
            true => self.make_block_focused(state, block),
            false => block,
        }
    }

    fn make_block_focused<'a>(&self, state: &State, block: Block<'a>) -> Block<'a> {
        block
            .border_style(Style::default().fg(state.theme.focused_border))
            .border_type(BorderType::Thick)
            .title_style(Style::default().bold())
    }

    fn init(&mut self) -> Result<(), TuiError> {
        Ok(())
    }

    fn handle_events(&mut self, event: Option<Event>) -> Result<Option<Action>, TuiError> {
        let r = match event {
            Some(Event::Key(key_event)) => self.handle_key_events(key_event)?,
            Some(Event::Mouse(mouse_event)) => self.handle_mouse_events(mouse_event)?,
            _ => None,
        };
        Ok(r)
    }

    fn handle_key_events(&mut self, _key: KeyEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn handle_mouse_events(&mut self, _mouse: MouseEvent) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn update(&mut self, _action: Action) -> Result<Option<Action>, TuiError> {
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError>;

    fn shortcuts(&self) -> Vec<Shortcut> {
        vec![]
    }
}
