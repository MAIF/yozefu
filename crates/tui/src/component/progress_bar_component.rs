//! Progress bar at the top of the window when you consume kafka records.
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Style},
    widgets::Widget,
};

#[derive(Clone, Default)]
pub(crate) struct ProgressBarComponent {
    length: u64,
    progress: u64,
}

impl ProgressBarComponent {
    pub fn new(length: u64) -> Self {
        Self {
            length,
            progress: 0,
        }
    }

    pub fn set_progress(&mut self, inc: usize) {
        self.progress = inc as u64;
    }

    pub fn set_length(&mut self, length: usize) {
        self.length = length as u64;
    }
}

impl Widget for ProgressBarComponent {
    fn render(self, area: Rect, buf: &mut Buffer) {
        if self.progress == 0 || self.length == 0 {
            return;
        }
        let percent = 100 * self.progress / self.length;
        let dd = u64::from(area.right() - area.left()) * percent / 100;
        buf.set_string(
            area.left(),
            area.top(),
            (0..dd).map(|_| "â–”").collect::<String>(),
            Style::default().fg(Color::Green),
        );
    }
}

#[test]
fn test_draw() {
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    let mut component = ProgressBarComponent::default();
    let mut terminal = Terminal::new(TestBackend::new(50, 3)).unwrap();
    component.set_length(100);
    component.set_progress(30);
    terminal
        .draw(|frame| {
            frame.render_widget(component, frame.area());
        })
        .unwrap();
    assert_snapshot!(terminal.backend());
}
