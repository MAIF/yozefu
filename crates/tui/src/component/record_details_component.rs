//! Component showing all the details of a given kafka record.
use bytesize::ByteSize;
use crossterm::event::{KeyCode, KeyEvent};

use itertools::Itertools;
use lib::{ExportedKafkaRecord, KafkaRecord};
use ratatui::{
    layout::Rect,
    style::{Style, Stylize},
    text::{Line, Span, Text},
    widgets::{Block, Borders, Clear, Padding, Paragraph, Wrap},
    Frame,
};
use tokio::sync::mpsc::UnboundedSender;

use crate::{error::TuiError, Action};

use super::{scroll_state::ScrollState, Component, ComponentName, Shortcut, State};

#[derive(Default)]
pub(crate) struct RecordDetailsComponent<'a> {
    record: Option<KafkaRecord>,
    lines: Vec<Line<'a>>,
    search_query: String,
    scroll: ScrollState,
    action_tx: Option<UnboundedSender<Action>>,
}

impl RecordDetailsComponent<'_> {
    pub fn new(_state: &State) -> Self {
        Self {
            ..Default::default()
        }
    }

    fn generate_span(key: &str, value: String) -> Line<'_> {
        Line::from(vec![
            Span::styled(
                format!("{:>12}: ", key.to_string()),
                Style::default().bold(),
            ),
            Span::styled(value.to_string(), Style::default()),
        ])
    }

    fn show_schema(&mut self) -> Result<(), TuiError> {
        if self.record.as_ref().map(|r| r.has_schemas()) == Some(false) {
            return Ok(());
        };

        let r = self.record.as_ref().unwrap();

        self.action_tx
            .as_ref()
            .unwrap()
            .send(Action::RequestSchemasOf(
                r.key_schema.as_ref().map(|s| s.id.clone()),
                r.value_schema.as_ref().map(|s| s.id.clone()),
            ))?;

        self.action_tx
            .as_ref()
            .unwrap()
            .send(Action::NewView(ComponentName::Schemas))?;
        Ok(())
    }

    fn compute_record_rendering(&mut self) {
        if self.record.is_none() {
            self.record = Some(KafkaRecord::default());
        }

        let record = self.record.as_ref().unwrap();
        let mut to_render = vec![
            Line::default(),
            Self::generate_span("Topic", record.topic.clone()),
            Self::generate_span(
                "Timestamp",
                self.record
                    .as_ref()
                    .unwrap()
                    .timestamp
                    .unwrap_or(0)
                    .to_string(),
            ),
            Self::generate_span(
                "DateTime",
                self.record
                    .as_ref()
                    .unwrap()
                    .timestamp_as_local_date_time()
                    .map(|e| e.to_rfc3339_opts(chrono::SecondsFormat::Millis, false))
                    .unwrap_or("".to_string()),
            ),
            Self::generate_span("Offset", record.offset.to_string()),
            Self::generate_span("Partition", record.partition.to_string()),
            Self::generate_span("Size", ByteSize(record.size as u64).to_string()),
            Self::generate_span("Headers", "".to_string()),
        ];

        let longest_header_key = self
            .record
            .as_ref()
            .unwrap()
            .headers
            .keys()
            .map(|e| e.len())
            .max()
            .unwrap_or(0);

        let mut formatted_headers = vec![];
        for entry in self
            .record
            .as_ref()
            .unwrap()
            .headers
            .iter()
            .sorted_by(|a, b| a.0.cmp(b.0))
            .enumerate()
        {
            let e = entry.1;
            match entry.0 {
                0 => formatted_headers.push(Span::styled(
                    format!("{: <width$}", e.0, width = longest_header_key),
                    Style::default().italic(),
                )),
                _ => formatted_headers.push(Span::styled(
                    format!("              {: <width$}", e.0, width = longest_header_key),
                    Style::default().italic(),
                )),
            };
            formatted_headers.push(Span::styled(" : ", Style::default()));
            formatted_headers.push(Span::styled(e.1.to_string(), Style::default()));
        }

        if !formatted_headers.is_empty() {
            let first = formatted_headers.iter().take(3).collect_vec();
            let line: &mut Line<'_> = to_render.last_mut().unwrap();
            line.spans.remove(1);
            for span in first {
                line.push_span(span.clone());
            }
        }

        for ppp in &formatted_headers.into_iter().skip(3).chunks(3) {
            to_render.push(Line::from(ppp.collect_vec()));
        }

        if let Some(s) = &record.key_schema {
            match &s.schema_type {
                Some(t) => to_render.push(Self::generate_span(
                    "Key schema",
                    format!("{} - {}", s.id, t),
                )),
                None => to_render.push(Self::generate_span("Key schema", s.id.to_string())),
            }
        }
        if let Some(s) = &record.value_schema {
            match &s.schema_type {
                Some(t) => to_render.push(Self::generate_span(
                    "Value schema",
                    format!("{} - {}", s.id, t),
                )),
                None => to_render.push(Self::generate_span("Value schema", s.id.to_string())),
            }
        }

        to_render.extend(vec![
            Self::generate_span("Key", record.key_as_string.clone()),
            Self::generate_span("Value", "".to_string()),
        ]);

        //let syntax = SYNTAX_SET.find_syntax_by_extension("json").unwrap();
        //let mut h = HighlightLines::new(syntax, &self.theme);
        //
        //let mut payload_lines = vec![];
        //for line in LinesWithEndings::from(&value_s) {
        //    let ranges: Vec<(syntect::highlighting::Style, &str)> =
        //        h.highlight_line(line, &SYNTAX_SET).unwrap();
        //    let escaped = as_24_bit_terminal_escaped(&ranges[..], false);
        //    let output = escaped.into_text().unwrap();
        //    payload_lines.extend(output.lines);
        //}
        let text = Text::from(record.value.to_string_pretty());
        to_render.extend(text.lines);

        self.lines = to_render;
        self.scroll.reset();
    }
}

impl Component for RecordDetailsComponent<'_> {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) -> Result<(), TuiError> {
        self.action_tx = Some(tx);
        Ok(())
    }

    fn id(&self) -> ComponentName {
        ComponentName::RecordDetails
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        match key.code {
            KeyCode::Char('k') => {
                self.scroll.scroll_to_next_line();
            }
            KeyCode::Char('j') => {
                self.scroll.scroll_to_previous_line();
            }
            KeyCode::Char('[') => {
                self.scroll.scroll_to_top();
            }
            KeyCode::Char(']') => {
                self.scroll.scroll_to_bottom();
            }
            KeyCode::Char('o') => {
                if let Some(record) = &self.record {
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::Open(record.clone()))?;
                }
            }
            KeyCode::Char('s') => self.show_schema()?,
            KeyCode::Char('c') => {
                if let Some(record) = &self.record {
                    let mut exported_record: ExportedKafkaRecord = record.into();
                    exported_record.search_query = self.search_query.to_string();
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::CopyToClipboard(
                            serde_json::to_string_pretty(&exported_record)
                                .expect("Unable to serialize record as json for the clipboard"),
                        ))?;
                }
            }
            KeyCode::Char('e') => {
                if let Some(record) = &self.record {
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::Export(record.clone()))?;
                }
            }
            _ => (),
        }
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::ShowRecord(record) => {
                self.record = Some(record);
                self.compute_record_rendering();
            }
            Action::Search(e) => self.search_query = e.query().to_string(),
            _ => {}
        };
        Ok(None)
    }

    fn shortcuts(&self) -> Vec<Shortcut> {
        let mut shortcuts = vec![
            Shortcut::new("J/K", "Scroll"),
            Shortcut::new("↑↓", "Prev/next record"),
        ];

        if self
            .record
            .as_ref()
            .map(|r| r.key_schema.is_some() || r.value_schema.is_some())
            .unwrap_or(false)
        {
            shortcuts.push(Shortcut::new("S", "Schemas"))
        }

        shortcuts
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let p = Paragraph::new(self.lines.clone())
            .wrap(Wrap { trim: false })
            .scroll((self.scroll.value(), 0));

        f.render_widget(Clear, rect);
        let block = Block::new()
            .borders(Borders::ALL)
            .padding(Padding::symmetric(4, 0))
            .title(" Record details ");
        let block = self.make_block_focused_with_state(state, block);

        f.render_widget(p.block(block), rect);
        self.scroll.draw(f, rect, self.lines.len() + 2);
        Ok(())
    }
}
