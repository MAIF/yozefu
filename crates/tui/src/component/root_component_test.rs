use crate::assert_draw;
use crate::component::Component;
use crate::{
    component::{ConcurrentRecordsBuffer, RootComponent, default_global_config, default_state},
    records_buffer::RecordsBuffer,
};
use std::sync::{Arc, LazyLock, Mutex};

static BUFFER: ConcurrentRecordsBuffer =
    LazyLock::new(|| Arc::new(Mutex::new(RecordsBuffer::new())));

#[test]
fn test_draw() {
    BUFFER.lock().unwrap().reset();
    let config = default_global_config();
    let mut component = RootComponent::new(
        "from begin".to_string(),
        vec!["topic1".to_string(), "topic2".to_string()],
        &config,
        &BUFFER,
        default_state(),
    );
    assert_draw!(component, 120, 20)
}
