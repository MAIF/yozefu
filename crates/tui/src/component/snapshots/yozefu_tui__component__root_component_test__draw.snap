---
source: crates/tui/src/component/root_component_test.rs
expression: terminal.backend()
---
"                                                                                                                        "
"  localhost   Topics                                                                                                   "
"                                                                                                                        "
" ┏ Topics [2] ━━━━━━━━━━━━━━━━━━━━┓ ╭ Records ────────────────────────────────────────────────────────────────────────╮ "
" ┃ [x] topic1                     ┃ │Timestamp                             Topic   Offset         Key  Value          │ "
" ┃ [x] topic2                     ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┃                                ┃ │                                                                                 │ "
" ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛ ╰─────────────────────────────────────────────────────────────────────────────────╯ "
" ╭ Search ────────────────────────────────────────────────────────────────────────────────────────────────────────────╮ "
" │ from begin                                                                                                         │ "
" ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ "
"                                                                                                                        "
"                                                                                                                        "
